File: backend-microservice/src/RabbitMq/consumer/rabbitmqConsumer.ts
import * as amqp from 'amqplib';
import { queueConfig } from '../../config/rabbitmqConfig';

class RabbitMqConsumer {
  private connection: amqp.Connection | null = null;
  private channel: amqp.Channel | null = null;
  private url = queueConfig;

  async connect(): Promise<void> {
    this.connection = await amqp.connect(this.url);
    this.channel = await this.connection.createChannel();
  }

  async consume(
    queue: string,
    onMessage: (msg: amqp.ConsumeMessage | null) => void
  ): Promise<void> {
    if (!this.channel) {
      throw new Error('Channel is not initialized. Call connect() first.');
    }

    await this.channel.assertQueue(queue, { durable: true });
    this.channel.consume(queue, onMessage, { noAck: true });
  }

  async close(): Promise<void> {
    if (this.channel) {
      await this.channel.close();
    }
    if (this.connection) {
      await this.connection.close();
    }
  }

  consumeMessage(msg: amqp.ConsumeMessage | null): void {
    if (msg) {
      console.log('Received message:', msg.content.toString());
    }
  }
}

export default RabbitMqConsumer;

File: model-microservice/rabbitmq/consumer/rabbitmqConsumer.py
import pika

class RabbitMqConsumer:
    def __init__(self, url: str):
        self.url = url
        self.connection = None
        self.channel = None

    def connect(self):
        self.connection = pika.BlockingConnection(pika.URLParameters(self.url))
        self.channel = self.connection.channel()

    def consume(self, queue: str, on_message_callback):
        if not self.channel:
            raise Exception('Channel is not initialized. Call connect() first.')

        self.channel.queue_declare(queue=queue, durable=True)
        self.channel.basic_consume(queue=queue, on_message_callback=on_message_callback, auto_ack=True)

    def start_consuming(self):
        if not self.channel:
            raise Exception('Channel is not initialized. Call connect() first.')

        self.channel.start_consuming()

    def close(self):
        if self.channel:
            self.channel.close()
        if self.connection:
            self.connection.close()


